ci: draft a playbook (#684)
sourcegraph/handbook: content/engineering/incidents/playbooks/ci.md

# CI Playbook

> This page is maintained by the [DevX Team](../../engineering/enablement/dev-experience/index.md).

> Audience: any software engineer, no prior infrastructure knowlegde required.

This document sums up what to do in multiple scenarios that can block the CI.

See also: [Continuous Integration](https://docs.sourcegraph.com/dev/background-information/continuous_integration) in our product documentation.

## Prerequisites

In order to handle problems with the CI:

- Have access to the `sourcegraph-ci` _project_ on Google Cloud Platform.
  - See [#it-tech-ops](https://sourcegraph.slack.com/archives/C01CSS3TC75)
- Have the CLI `gcloud` tool installed and have authenticated yourself.
  - See [Gain access to the cluster](../../engineering/deployments/debugging/tutorial/index.md)
- Set the current Kubernetes context to the `default-buildkite` _cluster_ in the `sourcegraph-ci` _project_:
  - `gcloud config set project sourcegraph-ci`

>   - `gcloud container clusters get-credentials default-buildkite --zone us-central1-c`

Reviewer: I think this should be part of `Gain access to the cluster` (if it isn't already)

Author: That page does mention these instructions, just not for the CI cluster. I thought it would be useful to have this here, it makes this self-contained in some way. 

Still, I agree with you that it should be present on that page, I've added some content in there.

The CI is what enables us to feel confident when delivering our changes to our users, and is one of the key components enabling Sourcegraph to deliver quality software. While the DevX team is in charge of managing the CI as a tool, it is essential for every engineer to be able to unblock themselves if there is a problem in order be autonomous.

This page lists common failures scenarios and provide a step by step guide to get the CI back in an operational state.

## Scenarios

### Build has failed on the `main` branch

- Gravity: _minor_
- Impact: that commit won't be deployed on DogFood and `sourcegraph.com` until an ulterior build passes.
- Possible causes:
  - The `main` branch runs additional checks compared to Pull Requests builds. So it's possible that one of those checks failed.
  - The `main` branch have changes that weren't in the Pull Request branch and those changes are causing a failure.
  - The `main` branch is failing due to a previous build.

#### Actions

1. Check your build on [Buildkite](https://buildkite.com/sourcegraph/sourcegraph/builds?branch=main)

- Find its link directly in the [#buildkite-main](https://sourcegraph.slack.com/archives/C02FLQDD3TQ) channel.

> - :bulb: Or run `sg ci status` in your shell, with the same branch checked out.

Reviewer: There's a few mentions of "the same branch" here, maybe we should just keep it as `main` for simplicity, or even give `--branch=main`?

2. Search for the failing steps, and browse the logs (:bulb: run `sg ci logs` in your shell, with the same branch checked out) .

- Look for a failure explanation: it can be a test that failed or a command that return a non zero exit code.

> 3. Check the previous builds on the `main` branch on [Buildkite](https://buildkite.com/sourcegraph/sourcegraph/builds?branch=main)

Reviewer: We should just refer to Loki here, perhaps an example query to demonstrate searching for a message

1. Are they failing with the same exact error?

   - Yes: see the [Builds are failing in the `main` branch with the same error]()
   - No: see next point.

1. Is that a real failure or a flake?

> 1. Restart that step. Maybe it will fail again, but if it doesn't it'll save you time.

Reviewer: Are these two steps related? (might be formatting)

   1. Revert the incriminating Pull Request.
   1. Checkout the PR branch
   1. Rebase it so it includes the changes that broke it when merged in the `main` branch.
   1. Rename the branch to `main-dry-run/your-branch` in order to get the CI to run the same exact checks it does on the `main` branch.
      c. No, but it seems to fail in step or code from another team.
   1. Reach out a member of the team responsible for that test.
   1. go for a. or b. from the previous points.
      d. No, and there is suspicion of a flake.
   - Yes: that's a flake. See the [Spotted a flake scenario]()

### Build are all failing on the `main` branch with the same error

- Gravity: _major_
- Impact: no commits are being deployed on DogFood and `sourcegraph.com` until the problem is resolved. Cutting a release is impossible.
- Possible causes:
  - A previous Pull Request introduced a change that causes a test to fail.
  - A previous Pull Request introduced a change that modified state in an unexpected way and broke the CI.
  - An external dependency is not available anymore and is causing builds to fail.
  - Some rate limiting API is throttling us and causing builds to fail.

#### Actions

> 1. Identify the error in common with the recent builds on [Buildkite](https://buildkite.com/sourcegraph/sourcegraph/builds?branch=main).

Reviewer: Also search Loki

1. Find the build where the problem appeared for the first time.

- :bulb: Often it's the first build that became red, but check that the error is the same to be sure.

1. Is this an external failure or an internal one?

- :bulb: External failures are about downloading a dependency like a package in a script or a in a Dockerfile. Often they'll manifest in the form of an HTTP error.
- :bulb: If unsure, ask for help on [#dev-chat](https://sourcegraph.slack.com/archives/C07KZF47K).
- Yes, it's an external failure:
  1. SSH into an agent to reproduce the problem:
  1. Launch `k9s` and select an agent, they are named like `buildkite-agent-xxxxxxxxxx-yyyyy`
  1. Press `s` to SSH on it.

>   1. You will be prompted with which pod to SSH onto, select the agent itself.

Reviewer: This could be a separate playbook entry I think (direct access to agent)

### Build are failing on the `main` branch with different errors

- Gravity: _major_
- Impact: no commits are being deployed on DogFood and `sourcegraph.com` until the problem is resolved. Cutting a release is impossible.
- Possible causes:
  - A previous Pull Request introduced a change that causes a test to fail.
  - A previous Pull Request introduced a change that modified state in an unexpected way and broke the CI.
  - An external dependency is not available anymore and is causing builds to fail under certain conditions.
  - Some rate limiting API is throttling us and causing builds to fail.

#### Actions

1. Escalate by creating an incident (`/incident` on Slack).
1. Get some help.
1. Downscale the agents to be able to observe exactly what's going on.
1. Update the `autoscaler` manifest to run a single agent
1. `cd sourcegraph/infrastructure`
1. Edit the [manifest](https://sourcegraph.com/github.com/sourcegraph/infrastructure/-/blob/buildkite/kubernetes/buildkite-autoscaler/buildkite-autoscaler.Deployment.yaml?L32-35) to have 1 on both max an min agent count.
1. `cd buildkite/kubernetes/buildkite-autoscaler`

> 1. Run `kubectl apply -n buildkite -f buildkite-autoscaler.Deployment.yaml`

Reviewer: ALso think this should be a standalone guide for making change

See also: [Continuous Integration](https://docs.sourcegraph.com/dev/background-information/continuous_integration) in our product documentation.

## Prerequisites

In order to handle problems with the CI, the following elements are necessary:

- Have access to the `sourcegraph-ci` _project_ on Google Cloud Platform.
  - See [#it-tech-ops](https://sourcegraph.slack.com/archives/C01CSS3TC75)
- Have the CLI `gcloud` tool installed and have authenticated yourself.
  - See [Gain access to the cluster](../../deployments/debugging/tutorial/index.md)
- (Optional) Install [K9s](https://k9scli.io) for easier interactions with the _pods_.

## Overview

The CI is what enables us to feel confident when delivering our changes to our users, and is one of the key components enabling Sourcegraph to deliver quality software. While the DevX team is in charge of managing the CI as a tool, it is essential for every engineer to be able to unblock themselves if there is a problem in order be autonomous.

This page lists common failures scenarios and provide a step by step guide to get the CI back in an operational state.

## Scenarios

### Build has failed on the `main` branch

- Gravity: _minor_

> - Impact: that commit won't be deployed on DogFood and `sourcegraph.com` until an ulterior build passes.

Reviewer: ```suggestion
- Impact: that commit won't be deployed on `k8s.sgdev.org` and `sourcegraph.com` until a later build passes.
```

- Possible causes:

>   - The `main` branch runs additional checks compared to Pull Requests builds. So it's possible that one of those checks failed.

Reviewer: Should note + link to stuff about how our pipeline is generated based on a variety of factors

1. Go the the "Explore" section
1. Make sure to select `grafanacloud-sourcegraph-logs` in the dropdown at the top of page.
1. Scope the time window to `7 Days` to make sure to find previous occurrences if there are any
1. Enter a query such as `{app="buildkite"} |= "your error message"` where "your error message" is a string that identiy approximately the failure cause observed in the failing step.
1. Is there a build that failed exactly like this?
   - **Yes**:
     1. 💡 Double check that you're looking at that the same step by inspecting the labels of message (click on the line to make them visible)
     1. **Yes**, that's a flake. See the [Spotted a flake scenario](#spotted-a-flake)
   - **No**: it's not a flake, reach out the team owning those tests.

### Restarting the agents

- Gravity: _minor_
- Impact: May fail ongoing builds, but that's fine.
- Possible causes:
  - Manual restart by an engineer.
  - Newer version of the agents needs to be deployed.

#### Actions

> 1. Open `k9s` to observe the currently running agents.

Reviewer: Are we recommending `k9s` for everyone? (I personally have never tried it before)

Author: I don't think that suggests that this is the only way to go.

Either you know your way with `kubectl` and you'll grasp `k9s` in like two minutes or you don't, in which case, `k9s` will make it much simpler to interact with the cluster.

Reviewer 2: I would suggest to use `kubectl get pods -w`  
Let's keep our playbook docs fairly generic 
```suggestion
1. Use`kubectl get pods -w` to observe the currently running agents (`k9s` works here too).
```

1. Is there a build that failed exactly like this?
   - **Yes**:
     1. 💡 Double check that you're looking at that the same step by inspecting the labels of message (click on the line to make them visible)
     1. **Yes**, that's a flake. See the [Spotted a flake scenario](#spotted-a-flake)
   - **No**: it's not a flake, reach out the team owning those tests.

### Restarting the agents

- Gravity: _minor_
- Impact: May fail ongoing builds, but that's fine.
- Possible causes:
  - Manual restart by an engineer.
  - Newer version of the agents needs to be deployed.

#### Actions

1. Open `k9s` to observe the currently running agents.
1. In a different terminal, run `kubectl -n buildkite rollout restart deployment buildkite-agent`.
1. Wait a bit to see the agents restarting completely.
1. Restart the faulty build and observe it the problem is fixed or not.
   - If necessary: escalate by creating an incident (`/incident` on Slack).

### SSH'ing into an agent

- Gravity: none
- Impact: none (unless a destructive action is performed)
- Possible cause:
  - Need to investigate a problem and suspect the agent is at fault

#### Actions

1. SSH into an agent to reproduce the problem:

> 1. Launch `k9s` and select an agent, they are named like `buildkite-agent-xxxxxxxxxx-yyyyy`
> 1. Press `s` to SSH on it.

Reviewer 2: Same as above, I would suggest `kubectl exec -it` here
```suggestion
1. Use `kubectl exec -it buildkite-agent-xxxx-yyyy` to open a shell 
```
Also, `kubectl exec` is not the same as SSH 😉 https://goteleport.com/blog/ssh-vs-kubectl/

> 1. You will be prompted with which pod to SSH onto, select the agent itself.

Reviewer 2: ```suggestion
1. You will be prompted with which pod to `exec` onto, select the agent itself.
```
